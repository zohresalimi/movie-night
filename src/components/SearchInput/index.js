import React, { useState, useContext, useEffect } from "react";
import { Input, Segment } from "semantic-ui-react";
// import axios from "axios";
import axiosInstance from "../../services";
import AppContext from "../../store/context";

import { SET_MOVIES_REDUCER, EMPTY_SEARCH_RESULT } from "../../constants";

import "./style.css";
const baseURL = process.env.REACT_APP_BASE_URL;

export default function SearchInput() {
  const { dispatch } = useContext(AppContext);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);
  const [message, setMessage] = useState("");
  const [searchTerm, setSearchTerm] = useState("");

  useEffect(() => {
    let mounted = true;
    let fetchTimeout;
    const fetchData = async () => {
      console.log("fetch data >>>>>>>>>>>>>>>>>");

      try {
        setIsLoading(true);
        const result = await axiosInstance.get(baseURL, {
          params: { query: searchTerm },
        });
        if (result && mounted) {
          dispatch({ type: SET_MOVIES_REDUCER, data: result.data });
        }
        if (!result.data.total_results) {
          setMessage("No movies found. Try with another term");
        }
      } catch (err) {
        setIsError(true);
      } finally {
        setIsLoading(false);
      }
    };

    if (searchTerm) {
      fetchTimeout = setTimeout(fetchData, 400);
    }

    return () => {
      clearTimeout(fetchTimeout);
      mounted = false;
    };
  }, [searchTerm, dispatch]);

  const handelInputChange = (value) => {
    setSearchTerm(value);
    if (!value) {
      dispatch({ type: EMPTY_SEARCH_RESULT });
    }
  };
  return (
    <Segment basic style={{ width: "100%" }}>
      <Input
        icon="search"
        loading={isLoading}
        data-testid="search-input"
        placeholder="Search by movie name!"
        value={searchTerm}
        onChange={(e) => handelInputChange(e.target.value)}
        className="full-width"
      />
      {isError && <div>Something went wrong ...</div>}
      {message && <div>{message}</div>}
      {isLoading && <div data-testid="loading">Loading...</div>}
    </Segment>
  );
}
