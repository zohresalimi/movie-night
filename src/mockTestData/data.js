import React, { useReducer } from "react";
import AppContext from "../store/context";
import reducer from "../store/reducer";

export const getTestState = () => {
  return {
    activePage: 1,
    movies: {
      1: {
        page: 1,
        total_pages: 2,
        total_results: 2,
        results: [
          {
            id: 10523,
            adult: false,
            poster_path: "/nJm8UF4zbUYEoku9seEtz1PTJGS.jpg",
          },
          {
            id: 50647,
            adult: false,
            poster_path: "/PlzkOA19mxEit7tV3AUnBYZ0io.jpg",
          },
        ],
      },
    },
    watchLaterList: {
      10523: {
        id: 10523,
        adult: false,
        poster_path: "/nJm8UF4zbUYEoku9seEtz1PTJGS.jpg",
      },
    },
    favoriteList: {
      10523: {
        id: 10523,
        adult: false,
        poster_path: "/nJm8UF4zbUYEoku9seEtz1PTJGS.jpg",
      },
    },
    apiConfig: {
      secure_base_url: "https://image.tmdb.org/t/p/",
      poster_sizes: ["w92", "w154", "w185", "w342", "w500", "w780"],
    },
  };
};

export const WithProvider = (props) => {
  const [state, dispatch] = useReducer(
    reducer,
    props.defaultValue || getTestState()
  );

  return (
    <AppContext.Provider
      value={{
        state,
        dispatch: props.mockDispatch || dispatch,
      }}
    >
      {props.children}
    </AppContext.Provider>
  );
};
