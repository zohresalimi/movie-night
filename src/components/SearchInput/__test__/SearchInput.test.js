import React from "react";
import { render, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import { WithProvider } from "../../../mockTestData/data";
import SearchInput from "../";

async function renderWrapper(props) {
  let component;

  act(() => {
    component = render(
      <WithProvider {...props}>
        <SearchInput />
      </WithProvider>
    );
  });

  return component;
}

describe("SearchInput Component Testing", () => {
  test("take snapshot", async () => {
    const { container } = await renderWrapper();
    expect(container).toMatchSnapshot();
  });
});
