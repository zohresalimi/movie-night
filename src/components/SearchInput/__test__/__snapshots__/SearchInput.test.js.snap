// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput Component Testing take snapshot 1`] = `
<div>
  <div
    class="ui basic segment"
    style="width: 100%;"
  >
    <div
      class="ui icon input full-width"
      data-testid="search-input"
    >
      <input
        placeholder="Search by movie name!"
        type="text"
        value=""
      />
      <i
        aria-hidden="true"
        class="search icon"
      />
    </div>
    
  </div>
</div>
`;
